[BEAM-2357]HiveIO implementation using native API - HCatalogReaderWriter

seshadri-cr opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

seshadri-cr commented:
Noting down initial comments from email thread - 

@jkff  -
From the point of view of general source/sink development, this code looks reasonable, except for a few violations of https://beam.apache.org/contribute/ptransform-style-guide/ (mainly around
https://beam.apache.org/contribute/ptransform-style-guide/#runtime-errors-and-data-consistency)
and other easily fixable things.

Would be good to get some extra input from people experienced with Hive on whether this is the right API or if there are any pitfalls to avoid.

@jbonofre 
- javadoc is not correct in HiveIO (it says write() for read ;)).
- estimated size is global to the table (doesn't consider the filter). It's not a big deal, but it should be documented.
- you don't use the desired bundle size provided by the runner for the split. 
You are using the Hive split count, which is fine, just explain in the main javadoc maybe.
- the reader should set current to null with nothing is read
- getCurrent() should throw NoSuchElementException in case of current is null
- in the writer, the flush should happen at the end of the batch as you did, but also when the bundle is finished

@iemejia 
maybe it is a good idea for the guys working on the Beam SQL implementation to take a look at their needs for this IO considering that it could be quite useful to test the SQL (given the structured nature of HCatalog).

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HCatRecordSimpleCoder.java:
```diff
@@ -0,0 +1,82 @@
+  @Override
+  public void encode(HCatRecord value, OutputStream outputStream)
+          throws IOException {
+    //TODO: Use HCatRecordSerDe to do a generic encode
```
Is it feasible to do this in the current PR?

seshadri-cr replied:
I guess this HCatRecordSimpleCoder is an overkill, I am  better off utilizing WritableCoder. I will change this !

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HCatRecordSimpleCoder.java:
```diff
@@ -0,0 +1,82 @@
+class HCatRecordSimpleCoder extends AtomicCoder<HCatRecord> implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(HCatRecordSimpleCoder.class);
+  private static final HCatRecordSimpleCoder INSTANCE = new HCatRecordSimpleCoder();
+  private static String metastoreUri;
```
This variable is unused, remove it

seshadri-cr replied:
Getting rid of this class, will utilize WritableCoder instead

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HCatRecordSimpleCoder.java:
```diff
@@ -0,0 +1,82 @@
+ * treating them as space delimited strings.
+ */
+class HCatRecordSimpleCoder extends AtomicCoder<HCatRecord> implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(HCatRecordSimpleCoder.class);
```
This logger is unused, remove it

seshadri-cr replied:
Getting rid of this class, will utilize WritableCoder instead

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HCatRecordSimpleCoder.java:
```diff
@@ -0,0 +1,82 @@
+  public HCatRecord decode(InputStream inputStream)
+      throws IOException {
+    StringBuilder content = new StringBuilder();
+    byte[] buffer = new byte[1024];
```
* Use standard utilities for reading from input streams, e.g. use StreamUtils
* Seems like you're consuming the entire remainder of the stream, but a Coder can be applied to a stream where there are other values encoded after this one.
If you're going through a String, then it's better to use a StringUtf8Coder instead of rolling your own encoding/decoding, e.g.:

```
private static final Coder<String> STRING_CODER = StringUtf8Coder.of();
...
STRING_CODER.encode(value.toString(), outputStream);
...
String value = STRING_CODER.decode(inputStream);
...

seshadri-cr replied:
Getting rid of this class, will utilize WritableCoder instead

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HCatRecordSimpleCoder.java:
```diff
@@ -0,0 +1,82 @@
+    while ((cnt = inputStream.read(buffer)) > 0) {
+      content.append(new String(buffer));
+    }
+    List<Object> objList = new ArrayList<Object>(Arrays.asList(content.toString().split(" ")));
```
Seems like this will corrupt the data if the string representation of any field in the record contains a space, is this correct?

It's probably necessary to do the right thing and use HCatRecordSerDe, instead of going through a string here.

seshadri-cr replied:
Getting rid of this class, will utilize WritableCoder instead

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+ * <p>Hive source supports reading of HCatRecord from a Hive instance.</p>
+ *
+ * <p>To configure a Hive source, you must specify a metastoreUri and a table name.
+ * Other optional parameters are database & filter
```
Does this javadoc render properly? I think `&` is a forbidden character.

seshadri-cr replied:
Replaced & with &amp;

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+      abstract Read build();
+    }
+
+    public Read withMetastoreUri(String metastoreUri) {
```
Please put javadocs on all public methods

seshadri-cr replied:
Done !

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+    }
+
+    protected Read withSplitId(Integer splitId) {
+      checkArgument(splitId >= 0);
```
The error message should include the actual value

seshadri-cr replied:
Will pass error message with split id

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+      return toBuilder().setFilter(filter).build();
+    }
+
+    protected Read withSplitId(Integer splitId) {
```
Why not int - can it be null? If yes, mark it `@Nullable` and document what does it mean if you supply null

seshadri-cr replied:
Changing this to a private method as that is sufficient

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+      return toBuilder().setSplitId(splitId).build();
+    }
+
+    protected Read withContext(ReaderContext context) {
```
All the protected methods here can probably be private

seshadri-cr replied:
Agreed, done !

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws
+    IOException, NoSuchObjectException, TException {
+      //TODO: This is not really used, should be removed
```
What is "this" referring to? Can you remove it here?

seshadri-cr replied:
getEstimatedSizeBytes() is not really used in this implementation as I am going with the splits returned by Hive API.  I will make this method simply return 0L with proper doc / comment stating that this implementation does not use the total size / bundle size to calculate the number of splits.

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+      IMetaStoreClient client = HCatUtil.getHiveMetastoreClient(hiveConf);
+      Table table = HCatUtil.getTable(client, spec.getDatabase(),
+        spec.getTable());
+      client.close();
```
Use try-with-resources for closing things robustly

seshadri-cr replied:
This piece of code is not relevant any more, have removed it

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws
+    IOException, NoSuchObjectException, TException {
```
Why so many throws declarations? I think `getEstimatedSizeBytes` is only allowed to throw `IOException`, right?

seshadri-cr replied:
I have updated this implementation to simply return 0L as this is not used in calculating splits, I am using the splits returned by native API. Hence I have removed all the exceptions as well.

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+      //just going with the default splits returned by Hive
+      //unable to explicitly control the splits returned by Hive
+      List<BoundedSource<HCatRecord>> sources = new ArrayList<>();
+      ReaderContext cntxt = getReaderContext();
```
nit: "context" is a better name, throughout the PR

seshadri-cr replied:
Changed to readerContext / writerContext across all occurrence

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+    private ReaderContext getReaderContext() {
+      ReadEntity entity = new ReadEntity.Builder().withDatabase(spec.getDatabase())
+          .withTable(spec.getTable()).withFilter(spec.getFilter()).build();
+      Map<String, String> hiveConfig = new HashMap<String, String>();
```
Can use ImmutableMap.of()

seshadri-cr replied:
Agreed

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+      try {
+        cntxt = masterReader.prepareRead();
+        } catch (HCatException e) {
+        LOG.error("Exception while preparing for read", e);
```
In this case you will return null - is this correct? Seems like a better action is to simply rethrow the exception (or to not catch it), because there is no way to recover from it.

seshadri-cr replied:
I will remove the try-catch-log-throw sequence, will make the exception as is. Will change the method to ... throws HCatException

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+    private static class BoundedHiveReader extends BoundedSource.BoundedReader<HCatRecord> {
+      private final BoundedHiveSource source;
+      private HCatRecord current;
+      Iterator<HCatRecord> hcatIterator;
```
private?

seshadri-cr replied:
some typo / copy-paste, will remove it

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+          hcatIterator = reader.read();
+        } catch (HCatException e) {
+          LOG.error("Exception while reading", e);
+          return false;
```
This means that in case of an exception, your reader will say that it has finished reading all data. From the point of view of a user, this is silent data loss - if there's no error surfaced to the user (i.e. if the pipeline succeeded), they won't be looking at the debug logs and won't see this LOG.error.

See https://beam.apache.org/contribute/ptransform-style-guide/#runtime-errors-and-data-consistency

seshadri-cr replied:
Will remove try-catch-log-throw sequence and make the method .. throws HCatException

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+
+      @Override
+      public void close() {
+        //nothing to close/release
```
Interesting... I'm surprised that reading a result set does not involve any closeable resources.

seshadri-cr replied:
Yes, seems to be happening under the hood of the API.

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+
+      @Override
+      public void populateDisplayData(
+          org.apache.beam.sdk.transforms.display.DisplayData.Builder builder) {
```
Qualified import seems unnecessary

seshadri-cr replied:
Will correct to DisplayData.Builder

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+        super.populateDisplayData(builder);
+        builder.add(DisplayData.item("database", spec.getDatabase()));
+        builder.add(DisplayData.item("table", spec.getTable()));
+        builder.add(DisplayData.item("filter", (spec.getFilter() != null
```
String.valueOf(spec.getFilter())

seshadri-cr replied:
Agree

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+          writerContext = getWriterContext();
+          slaveWriter = DataTransferFactory.getHCatWriter(writerContext);
+        } catch (HCatException e) {
+          LOG.error("Exception in initiateWriterContext", e);
```
This is an unnecessary log/rethrow - it will be obvious from the stacktrace that the exception happened in `initiateWrite`.

seshadri-cr replied:
Will remove the try-catch-log-throw sequence and make the method .. throws HCatException

jkff commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java:
```diff
@@ -0,0 +1,474 @@
+
+      private void flush() throws HCatException {
+        try {
+          //write data using slaveWriter
```
These comments seem unnecessary because they don't communicate any information that is not already apparent from the code.

seshadri-cr replied:
Agree, will remove them

seshadri-cr commented:
@jkff, thanks for reviewing!  I will work on addressing them and move on to unit testing next.

seshadri-cr commented on sdks/java/io/hive/pom.xml:
```diff
@@ -0,0 +1,138 @@
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
```
I will format the POM file as expected

seshadri-cr commented on sdks/java/io/hive/pom.xml:
```diff
@@ -0,0 +1,138 @@
+
+   <properties>
+    <hive.version>2.1.1</hive.version>
+	<junit.version>4.12</junit.version>
```
Yes, I will remove this

seshadri-cr commented on sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HCatRecordSimpleCoder.java:
```diff
@@ -0,0 +1,82 @@
+ */
+package org.apache.beam.sdk.io.hive;
+
+//import java.io.DataOutput;
```
Getting rid of this class, will utilize WritableCoder instead

jkff commented:
Github's UI makes it difficult to track which comments have been responded to when I receive notifications - please post a comment on this PR explicitly saying when you'd like me to take another look (though responses individual comments are still useful to make sure they were all handled in some way).

seshadri-cr commented:
@jkff , I have addressed your comments + those of @jbonofre . I am waiting for a consensus on @ssisk 's comments on using GroupIntoBatchesTransform

jkff commented:
Code looks pretty good as-is (I can do remaining fixups myself when merging), but yeah, I agree with @ssisk that `GroupIntoBatches` would make it even simpler.

Looking forward to the tests - to clarify, expecting to have them in the current PR (we commit tests together with code so that users can always rely that code exposed to them is already tested)!

iemejia commented:
@ssisk and @jkff you are right about the fact that GroupIntoBatches is a good simplification for this code however I don't think we should do this at this moment, because there is still a big issue. Not all the runners support the State/Timer API today so this will limit the use of this IO to only the runners who do.

I think we should let create a new independent JIRA for the use of GroupIntoBatches as once most of the runners support both the State/Timer APIs. We can even create a meta-JIRA to cover other IOs that do batches like this, e.g. ElasticSearchIO, another idea would be to include this in a de-facto and abstract manner for the SDF API.

iemejia commented:
@seshadri-cr Another thing that we already discussed in the review of the HIFIO based version of the IO is that the Hive name is inappropriate considering the fact that this IO is used only to access the HCatalog part of Hive. So please don't forget to rename this in every place it makes sense (even for the module name).

iemejia commented:
@seshadri-cr Also remember to add <module>hcatalog</module> in the sdks/java/io/pom.xml file so the io includes the parent version dependencies.

seshadri-cr commented:
Yes, I will work on the unit test for this PR. I will update it once ready !

Regards,
Seshadri
408 601 7548

On May 24, 2017 9:29 PM, "Eugene Kirpichov" <notifications@github.com>
wrote:

> Code looks pretty good as-is (I can do remaining fixups myself when
> merging), but yeah, I agree with @ssisk <https://github.com/ssisk> that
> GroupIntoBatches would make it even simpler.
>
> Looking forward to the tests - to clarify, expecting to have them in the
> current PR (we commit tests together with code so that users can always
> rely that code exposed to them is already tested)!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/3220#issuecomment-303919464>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AauNhXqfARJ08XsK0xIoZ6QDSh3C9Cj2ks5r9QPAgaJpZM4NlYah>
> .
>

seshadri-cr commented:
Thanks for the reminder :-) , I ll change all reference to HCatalog plus
include in parent pom as well.

Regards,
Seshadri

On May 25, 2017 2:30 AM, "Ismaël Mejía" <notifications@github.com> wrote:

> @seshadri-cr <https://github.com/seshadri-cr> Another thing that we
> already discussed in the review of the HIFIO based version of the IO is
> that the Hive name is inappropriate considering the fact that this IO is
> used only to access the HCatalog part of Hive. So please don't forget to
> rename this in every place it makes sense (even for the module name).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/3220#issuecomment-303965842>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AauNhSsDZvTiMMYfMU5OImGsXjy3AZHfks5r9UoXgaJpZM4NlYah>
> .
>

iemejia commented:
@seshadri-cr For the unit test you can take a look at https://github.com/apache/hive/blob/master/hcatalog/core/src/test/java/org/apache/hive/hcatalog/data/TestReaderWriter.java
and the MiniCluster class they also have to see if you can create a test case with multiple slaves so you can also test the split.

seshadri-cr commented:
@jkff , I have added the unit testing for which I used an embedded source used in the link  https://github.com/apache/hive/blob/master/hcatalog/core/src/test/java/org/apache/hive/hcatalog/data/TestReaderWriter.java suggested by @iemejia .

I used SourceTestUtils for source related part and end-to-end testing for overall read & write scenarios.

Looking forward to your review & comments !

ssisk commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+          }
+        }));
+    PAssert.thatSingleton(output.apply("Count All", Count.<String>globally()))
+    .isEqualTo(100L);
```
please do a better check than just count.  PAssert.that(collection).containsInAnyOrder() will do what you want.

ssisk replied:
Also - note that eugene above pointed out some useful tools like SourceTestUtils

seshadri-cr replied:
Agree, I will change it to use containsInAnyOrder() for a better check.

ssisk commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+/**
+ *Test for HCatalogIO.
+ */
+public class HCatalogIOTest extends HCatBaseTest implements Serializable {
```
It looks like you pull in a property called hiveconf to connect to an external version of the data store.  That'd make this an integration test (like JdbcIOIT) rather than a unit test (like JdbcIOTest). As Ismael mentioned, you should look into minicluster as a way to connect to an in-memory version of the data store.

We have a guide that helps describe how tests are structured in beam - you can find it here: https://docs.google.com/document/d/153J9jPQhMCNi_eBzJfhAg-NprQ7vbf1jNVRgdqeEE8I/edit?usp=sharing - it's partially still in the works, so let me know if you have questions. 

If you want to check in an IT (which is great!) you'll need to check in a kubernetes script that sets up an instance of your store. You'll also need to receive the connection configuration via pipelineoptions, not maven options. JdbcIOIT has good examples of this.

seshadri-cr replied:
To clarify, the UT here does not require any additional scripts / setup apart from the check-ins.
 I have used a light-weight embedded Hive source - https://github.com/apache/hive/blob/0af6cb42725659740a022044c6cc464ef1cf4e6b/hcatalog/core/src/test/java/org/apache/hive/hcatalog/mapreduce/HCatBaseTest.java which brings up an instance of the metastore when the UT is executed. I felt this does not qualify as an IT as there is no persistent external Hive instance that is set-up, WDYT.

ssisk commented:
@iemejia thanks for letting me know that group into batches doesn't work on all runners. That's disappointing.

jkff commented on sdks/java/io/hcatalog/pom.xml:
```diff
@@ -0,0 +1,269 @@
+  <properties>
+    <hive.version>2.1.0</hive.version>
+
+	<!-- properties for test -->
```
Please use consistent indentation in pom.xml. I wonder if your IDE is configured in a way that it doesn't respect existing indentation?..

seshadri-cr replied:
Guess I mixed-up space & tab, will correct this !

jkff commented on sdks/java/io/hcatalog/pom.xml:
```diff
@@ -0,0 +1,269 @@
+	<jersey.version>1.14</jersey.version>
+    <test.log4j.scheme>file://</test.log4j.scheme>
+    <project.build.directory>${project.basedir}/target</project.build.directory>
+    <test.tmp.dir>${project.build.directory}/tmp</test.tmp.dir>
```
Seems icky, you should be able to use regular Java temp file creation mechanisms, as well as the JUnit `@Rule TemporaryFolder` (http://junit.org/junit4/javadoc/4.12/org/junit/rules/TemporaryFolder.html).

seshadri-cr replied:
I will remove these steps from pom and have them as a part of UT setup.

jkff commented on sdks/java/io/hcatalog/pom.xml:
```diff
@@ -0,0 +1,269 @@
+      <scope>provided</scope>
+    </dependency>
+
+	<dependency>
```
Really a lot of indentation is off in this file.

seshadri-cr replied:
Will correct it

jkff commented on sdks/java/io/hcatalog/resources/testconf/hive-log4j2.properties:
```diff
@@ -0,0 +1,123 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Is all of this configuration really necessary to bring up an in-memory instance of Hive for unit testing?

iemejia replied:
Same question as @jkff. I think this should be removed if not mandatory.

seshadri-cr replied:
I have refactored these and do not need hive-log4j2.properties anymore. The only required file is hive-site.xml which I will move inside test/resources.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,507 @@
+ *
+
+ */
+@Experimental
```
Why is this marked experimental?

seshadri-cr replied:
Ah ! looks to be a copy-paste, will remove it

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,507 @@
+
+/**
+ * IO to read and write data using HCatalog.
+ *
```
A lot of unnecessary blank lines throughout this whole javadoc comment.

seshadri-cr replied:
I will remove the unnecessary ones

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,507 @@
+    /**
+     * Sets the configuration properties like metastore URI.
+     * This is mandatory
+     * @param configProperties
```
These javadocs don't add value - it's already obvious from the method signature that there's a parameter called `configProperties`, and the `@return` tag is empty so it doesn't explain anything at all. Suggest to leave only the first two lines. Same cleanup applies to other javadoc comments.

seshadri-cr replied:
Agree, will correct as suggested

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,507 @@
+    }
+
+    /**
+     * This implementation simply returns 0L as this is not used in calculating the splits.
```
If this `BoundedSource` is unable to dynamically split, unable to estimate size, and unable to use `desiredBundleSizeBytes` when splitting, then it shouldn't be a `BoundedSource`, it should simply be a composition of two `DoFn`'s (with a shuffle in the middle to ensure parallelism) - see e.g. how `DatastoreIO` is implemented: there's one `DoFn` to split and another one to read.

`BoundedSource` is an advanced and error-prone API and should be used only when strictly necessary. See https://beam.apache.org/documentation/io/authoring-overview/#when-to-implement-using-the-source-api

seshadri-cr replied:
I think this is clarified now, will go ahead with BoundedSource as per our latest conversation

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,507 @@
+    /**
+     * Sets the configuration properties like metastore URI.
+     * This is mandatory
+     * @param metastoreUri
```
Same comment cleanups apply

seshadri-cr replied:
Agree, will correct as suggested

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+    createTestTable();
+    pipeline.apply(
+        Create.of(
+            toHCatRecord(0), toHCatRecord(1), toHCatRecord(2), toHCatRecord(3),
```
This should be a loop.

seshadri-cr replied:
Yes, I will make it a loop

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+    thrown.expectMessage(containsString("NoSuchObjectException"));
+    pipeline.apply(
+        Create.of(
+            toHCatRecord(0), toHCatRecord(1), toHCatRecord(2), toHCatRecord(3),
```
Same. Moreover, in this test the Create.of() data doesn't matter at all, because the test verifies failure to write it.

seshadri-cr replied:
Changed it as a loop, however Create.of() is required in this case as otherwise I cannot apply HCatalogIO.write() on the pipeline

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+    pipeline.apply(
+        Create.of(
+            toHCatRecord(0), toHCatRecord(1), toHCatRecord(2), toHCatRecord(3),
+            toHCatRecord(4), toHCatRecord(5), toHCatRecord(6), toHCatRecord(7),
```
Same

seshadri-cr replied:
Will remove TestPipeline for transform validations

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage(containsString("configProperties"));
+    pipeline.apply(
+        Create.of(
```
Same

Actually, for these tests that verify that the transform itself is malformed, you don't even need the `TestPipeline`. You can just do `HCatalogIO.write().withTable("myowntable").validate(null)`.

seshadri-cr replied:
Yes, will remove TestPipeline for transform validations

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+    pipeline.apply(HCatalogIO.read()
+        .withConfigProperties(map)
+        .withTable("myowntable"))
+        .apply(ParDo.<DefaultHCatRecord, String>of(new DoFn<DefaultHCatRecord, String>() {
```
Why do you need this ParDo in this test?

seshadri-cr replied:
Yes, it doesn't make sense here. I will remove it

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+    thrown.expectMessage(containsString("configProperties"));
+    pipeline.apply(HCatalogIO.read()
+        .withTable(TEST_TABLE_NAME))
+        .apply(ParDo.<DefaultHCatRecord, String>of(new DoFn<DefaultHCatRecord, String>() {
```
Same

seshadri-cr replied:
Will remove it

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+    Map<String, String> map = new HashMap<String, String>();
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage(containsString("table"));
+    pipeline.apply(HCatalogIO.read()
```
Same

seshadri-cr replied:
Will remove it

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+      recordsList.addAll(SourceTestUtils.readFromSource(sourceList.get(i),
+        PipelineOptionsFactory.create()));
+    }
+    Assert.assertTrue(recordsList.size() == 100);
```
This should be `assertEquals`

seshadri-cr replied:
Will change it to use Assert.assertThat(T, Matcher<T>)

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,342 @@
+  public void testSourcesEqualReferenceSource() throws Exception {
+    List<BoundedHCatalogSource> sourceList = getSourceList();
+    for (int i = 0; i < sourceList.size(); i++) {
+      SourceTestUtils.assertSourcesEqualReferenceSource(sourceList.get(i),
```
All `SourceTestUtils` tests will become obsolete after converting to `DoFn`.

seshadri-cr replied:
SourceTestUtils tests will still remain as we agreed to use BoundedSource

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTestUtils.java:
```diff
@@ -0,0 +1,136 @@
+
+  /**
+   * Returns the count of records read from the table.
+   * @param map
```
Same comments about unnecessary javadocs

seshadri-cr replied:
Will address them

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTestUtils.java:
```diff
@@ -0,0 +1,136 @@
+    */
+   public static void writeRecords(WriterContext context) throws HCatException {
+    HCatWriter writer = DataTransferFactory.getHCatWriter(context);
+    writer.write(new HCatRecordItr());
```
Seems better to just create a `List` of the expected records, and call `writer.write(records.iterator())` than rolling your own iterator class which is more code and error-prone.

seshadri-cr replied:
Agreed, will remove the iterator class

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTestUtils.java:
```diff
@@ -0,0 +1,136 @@
+    * @return
+    */
+   public static DefaultHCatRecord toHCatRecord(int value) {
+    List<Object> dataList = new ArrayList<Object>(2);
```
Can be condensed into one line: `return new DefaultHCatRecord(Arrays.asList("record " + value), value)`

seshadri-cr replied:
Agree

jkff commented on sdks/java/io/pom.xml:
```diff
@@ -71,6 +71,7 @@
     <module>hadoop-file-system</module>
     <module>hadoop</module>
     <module>hbase</module>
+	<module>hcatalog</module>
```
Indentation

seshadri-cr replied:
Will address this

seshadri-cr commented:
@jkff, regret the indentation issues on the pom file. There seems to be some mess-up, I find that these files are actually indented properly in my local git however when I committed these changes they did not reflect in my fork (which I did not realize until your comments). I would correct this and double check before I get this for review.

Also I have a few questions that would help me addressing the major comments :

**1. Use of BoundedSource -**
   I was looking into some more details on the aspect of splits and stumbled across this - https://cwiki.apache.org/confluence/display/Hive/HCatalog+Configuration+Properties. There seems to be a way to indicate HCatalog a 'desired' number of splits using the config parameter - 'hcat.desired.partition.num.splits'. However this is used as an 'hint' by HCatalog and does not guarantee that the exact number of desired splits will be achieved. Though this is a best-effort approach, does it make sense to implement this split logic and consider keeping it as a BoundedSource as this is the best attempt to get the desired number of splits. Else I can break this up as DoFns as you suggested.
Please let me know your thoughts on this.

**2. Lot of config for embedded Hive source**
  I completely agree with your comments and in fact felt the same way myself too. This is the only suitable  embedded Hive source that was available for the specific use-case. There are two config files in total viz. hive-site.xml and hive-log4j2.properties. I guess the log4j props can be removed however hive-site.xml contains the config params required to bring up the embedded Hive server, I will try to see if there is an alternate for this. In case if I do not find an alternate, it might have to stay on. @iemejia , please let me know if you have any pointers to avoid the usage of hive-site.xml

**3. Rest of the comments**
I agree with rest of your comments, I will address them as specified.

Appreciate all your efforts in reviewing this, being my first contribution here I am learning a lot from you folks and thankful for all your guidance and support in making this work !

jkff commented:
Thanks!

Re indentation: could it be a tabs/spaces mix-up?

Re BoundedSource: unfortunately currently the BoundedSource API takes a "desired split size in bytes" hint, rather than "number of splits" hint, so if HCatalog can only split into desired number of splits, that doesn't make BoundedSource a better candidate than DoFn. So DoFn is still preferable.

Re embedded Hive config: seems like these hive-site.xml and hive-log4j2.properties are exact copies of those in https://github.com/apache/hive/blob/master/data/conf/hive-site.xml ? I guess in that case it's ok to have the large configs, but it'd be good to add a comment in those files explaining where they come from.

Also: how long does the unit test take to run? I'm wondering, if it requires this much set up, it might be too heavyweight to be a unit test and perhaps should be changed into unit tests with mocks + integration tests with real Hive. I guess I'd set the bar at around 15-20 seconds.

seshadri-cr commented:
The BoundedSource implementation here would work with "desired split size in bytes" - hint 

- desiredSplits  = ( getEstimatedSizeBytes() / "desired split size in bytes" - hint passed by the API /  ) + 1
- getEstimatedSizeBytes() in this implementation currently returns 0, but can revert it back to return the table size instead (without the filter, if any passed) - https://github.com/seshadri-cr/beam/blob/b74523c13e03dc70038bc1e348ce270fbb3fd99b/sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java#L241
- set  'desiredSplits' as a param to HCatalog
Wondering if this is the expectation or if I am missing something.

On hive config, I agree to your suggestion to add the source of that file.

The UT currently takes around 18-19 seconds - WDYT.

[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.beam.sdk.io.hcatalog.HCatalogIOTest
[INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 18.641 s - in org.apache.beam.sdk.io.hcatalog.HCatalogIOTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0

jkff commented:
Oh I see. Then I guess setting desired splits in the way you propose makes sense, and keeping as BoundedSource makes sense.

19 seconds - fine to keep as a unit test then.

seshadri-cr commented:
@jkff , the code looks in a better shape for a review. 
A short summary of things I have addressed -
1. getEstimatedSizeBytes() and split() methods for BoundedSource

    https://github.com/seshadri-cr/beam/blob/0a0162820009a2e4b0a8e1209161b960b3cbd7b5/sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java#L263

https://github.com/seshadri-cr/beam/blob/0a0162820009a2e4b0a8e1209161b960b3cbd7b5/sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java#L289

2. Clean-up of UT setup
    I have removed the UT setup steps from pom file and also refactored UT code. The only config file required now is hive-site.xml

Apart from these I have addressed the code review comments on pom file indentation, java docs and other comments(I have individually replied for each comment)

seshadri-cr commented:
One observation, the UT took 32 seconds in the pre-commit Jenkins build. On my local machine it takes between 16-18 seconds.

seshadri-cr commented:
https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/12069/org.apache.beam$beam-sdks-java-io-hcatalog/testReport/org.apache.beam.sdk.io.hcatalog/HCatalogIOTest/

One of the SourceTestUtils test-case seems to be taking 16 seconds, https://github.com/seshadri-cr/beam/blob/37b586cdff924975507c1b4cc6802f23ee523570/sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java#L227

I am not quite sure why this takes such a long time, please let me know if you have any pointers

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,518 @@
+ * For instance:</p>
+ *
+ * <pre>{@code
+ * * Map<String, String> configProperties = new HashMap<String, String>();
```
extra star

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/EmbeddedMetastoreService.java:
```diff
@@ -0,0 +1,120 @@
+
+    //below properties are used by theembedded metastore to store config
+    //and data files during execution
+    System.setProperty("test.tmp.dir", HCatUtil.makePathASafeFileName(
```
This seems icky to do in a `static` block. What if another test also uses `test.tmp.dir`?

seshadri-cr replied:
Removed the use of static

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/EmbeddedMetastoreService.java:
```diff
@@ -0,0 +1,120 @@
+  private static void setUpFSDirectory(String dirPath) {
+    File fsDir = new File(dirPath);
+    fsDir.mkdir();
+    fsDir.deleteOnExit();
```
This definitely should be a `@Rule TemporaryFolder` then.

seshadri-cr replied:
Addressed in other comment

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,250 @@
+  public void testSourcesEqualReferenceSource() throws Exception {
+    List<BoundedHCatalogSource> sourceList = getSourceList();
+    for (int i = 0; i < sourceList.size(); i++) {
+      SourceTestUtils.assertSourcesEqualReferenceSource(sourceList.get(i),
```
`assertSourcesEqualReferenceSource` would be more appropriate to call on a top-level, unsplit source - but here you're calling it on results of splitting (produced in `getSourceList`). It's going to also be much faster because you'll call this function only once.

seshadri-cr replied:
Changed it as shown here - https://github.com/seshadri-cr/beam/blob/hiveio/sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java#L241

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTestUtils.java:
```diff
@@ -0,0 +1,194 @@
+   /**
+    * Returns a list of DefaultHCatRecords of passed size.
+    */
+   static List<DefaultHCatRecord> getDefaultHCatRecords(int size) {
```
Seems almost identical to `getHCatRecords`.

iemejia replied:
Agree with @jkff you can remove this method and rely on getHCatRecords (you should make it accesible) since the transform uses raw HCatRecord. Also you can make the method toHCatRecord return just HCatRecord.

seshadri-cr replied:
Got this out.

iemejia commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,250 @@
+   */
+  @Test
+  @TestDataInitializationReqd
+  public void testWriteWriteThenReadSuccess() throws Exception {
```
Remove the extra 'Write', rename to testWriteThenReadSuccess

seshadri-cr replied:
done !

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,518 @@
+   * A {@link PTransform} to write to a HCatalog managed source.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<DefaultHCatRecord>, PDone> {
```
Make the type here HCatRecord, better than DefaultHCatRecord

seshadri-cr replied:
done !

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,518 @@
+      checkNotNull(getTable(), "table");
+    }
+
+    private static class WriteFn extends DoFn<DefaultHCatRecord, Void> {
```
Make the type here HCatRecord, better than DefaultHCatRecord

seshadri-cr replied:
done !

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,518 @@
+      }
+
+      private void flush() throws HCatException {
+        try {
```
Add a validation to see if the collection is empty before trying to flush it, e.g.
```
        if (hCatRecordsBatch.isEmpty()) {
          return;
        }
```

seshadri-cr replied:
makes sense, done

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,518 @@
+   */
+  @VisibleForTesting
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<DefaultHCatRecord>> {
```
Can this be changed from DefaultHCatRecord into HCatRecord (and the same for the Source and Reader) ?

seshadri-cr replied:
Changed it now, there was some mismatch with the WritableCoder, it was trying to instantiate an instance of HCatRecord (which is the parameterized type) and threw an exception. Have handled it like this - https://github.com/seshadri-cr/beam/blob/hiveio/sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java#L243

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,518 @@
+ * }</pre>
+ */
+
+public class HCatalogIO {
```
Please add the @Experimental annotation.

seshadri-cr replied:
I had this earlier (though not intentionally, copy-paste from other transform) but then got a review comment from @jkff trying to know if there is any specific reason for marking this as Experimental for which I could not think of any specific reason and hence I removed this.
When should an IO transform be marked as Experimental, if this applies for all *new* transforms then  yes I can mark it back.

iemejia replied:
You should put it back, the Experimental annotation means that the IO public API is still evolving, it is a bit of an unfortunate name. I prefer that we put this annotation back because this is the first version of the IO and it is better not to engage with code stability too early, in case we find (or someone reports) new things that can be improved later on via breaking changes.
Please change it, I suppose @jkff somehow forgot this, but we have this annotation at this moment in almost all the IOs.

seshadri-cr commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,518 @@
+ * <h3>Reading using HCatalog</h3>
+ *
+ * <p>HCatalog source supports reading of DefaultHCatRecord from a HCatalog managed source,
+ * for eg. Hive.</p>
```
Addressed these.

seshadri-cr commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,517 @@
+      //pass the 'desired' split count as an hint to the API
+      configProps.put(HCatConstants.HCAT_DESIRED_PARTITION_NUM_SPLITS,
+          String.valueOf(desiredSplitCount));
+      configProps = Collections.unmodifiableMap(configProps);
```
Agreed, rather than making a copy here I made this while building the pipeline - https://github.com/seshadri-cr/beam/blob/hiveio/sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java#L172

seshadri-cr commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -0,0 +1,517 @@
+      configProps = Collections.unmodifiableMap(configProps);
+      HCatReader masterReader = DataTransferFactory.getHCatReader(entity,
+          spec.getConfigProperties());
+      ReaderContext readerContext = null;
```
done !

seshadri-cr commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/EmbeddedMetastoreService.java:
```diff
@@ -0,0 +1,120 @@
+public class EmbeddedMetastoreService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(EmbeddedMetastoreService.class);
+  private static final String BASE_DIR = System.getProperty("user.dir");
```
I have refactored this class avoiding the use of any static and used JUnit tmp folders. I found a couple of odd things while using JUnit temporary folders - 
**_1._**  I had to explicitly invoke create() method as shown here - https://github.com/seshadri-cr/beam/blob/hiveio/sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java#L119. Otherwise when I access the folder instance to pass on the path as shown here - https://github.com/seshadri-cr/beam/blob/hiveio/sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java#L120 it fails with IllegalStateException - the temporary folder has not yet been created.
**_2_**  Had to explicitly specify this - https://github.com/seshadri-cr/beam/blob/hiveio/sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/EmbeddedMetastoreService.java#L69 to create the workspace folders created under JUnit tmp folder. If the JUnit tmp folder is not empty at the end of the run, I noticed it does not delete it

seshadri-cr commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,250 @@
+   */
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target({ElementType.METHOD})
+  @interface TestDataSetupReqd {
```
done !

seshadri-cr commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -0,0 +1,250 @@
+
+    pipeline.run();
+
+    PCollection<String> output = pipeline.apply(HCatalogIO.read()
```
Interesting it worked with same pipeline! I have modified to use two separate pipelines in this case.

seshadri-cr commented on sdks/java/io/pom.xml:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="UTF-8"?>
+	<?xml version="1.0" encoding="UTF-8"?>
```
Oops, my bad ! Reverted it back

seshadri-cr commented:
@jkff, @iemejia  thanks again for taking time to review this. I have addressed these comments in the latest commit. 

PS:
I have tightened up by marking classes / members as final where applicable and also changed the methods to use default / private access modifier in certain places, though I did not use any static code analysis tool.

seshadri-cr commented:
Ok I ll add it back then.

Regards,
Seshadri

On Jun 7, 2017 11:10 PM, "Ismaël Mejía" <notifications@github.com> wrote:

> *@iemejia* commented on this pull request.
> ------------------------------
>
> In sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/
> io/hcatalog/HCatalogIO.java
> <https://github.com/apache/beam/pull/3220#discussion_r120806378>:
>
> > + * configProperties.put("hive.metastore.uris","thrift://metastore-host:port");
> + *
> + * pipeline
> + *   .apply(...)
> + *   .apply(HiveIO.write()
> + *       .withConfigProperties(configProperties) //mandatory
> + *       .withTable("employee") //mandatory
> + *       .withDatabase("default") //optional, assumes default if none specified
> + *       .withFilter(partitionValues) //optional,
> + *       should be specified if the table is partitioned
> + *       .withBatchSize(1024L)) //optional,
> + *       assumes a default batch size of 1024 if none specified
> + * }</pre>
> + */
> +
> +public class HCatalogIO {
>
> You should put it back, the Experimental annotation means that the IO
> public API is still evolving, it is a bit of an unfortunate name. I prefer
> that we put this annotation back because this is the first version of the
> IO and it is better not to engage with code stability too early, in case we
> find (or someone reports) new things that can be improved later on via
> breaking changes.
> Please change it, I suppose @jkff <https://github.com/jkff> somehow
> forgot this, but we have this annotation at this moment in almost all the
> IOs.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/3220#discussion_r120806378>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AauNhZFlrBanVLplXeSPea285Q2LEJTeks5sB5BcgaJpZM4NlYah>
> .
>

iemejia commented:
It is in a quite good shape now, don't worry about the Experimental I can add this later on as well as some small fixes I have in mind (to apply probably after the small ones Eugene mentioned if they are different).
@jkff What do you think, does it LGTY so we can merge it?

iemejia commented:
Mmm I just run `mvn clean verify -Prelease` and there is apparently a file that is generated during the tests that breaks the RAT validations, we should ensure that the test/data directory gets removed or it will break on this validation @seshadri-cr .

seshadri-cr commented:
Oh, it passed clean in my local. Let me try the build in a Unix box.

[INFO] --- apache-rat-plugin:0.12:check (default) @ beam-sdks-java-io-hcatalog ---
[INFO] Enabled default license matchers.
[INFO] Will parse SCM ignores for exclusions...
[INFO] Finished adding exclusions from SCM ignore files.
[INFO] 61 implicit excludes (use -debug for more details).
[INFO] Exclude: **/target/**/*
[INFO] Exclude: **/dependency-reduced-pom.xml
[INFO] Exclude: **/hs_err_pid*.log
[INFO] Exclude: .github/**/*
[INFO] Exclude: **/*.iml
[INFO] Exclude: **/.idea/**/*
[INFO] Exclude: **/*.egg-info/**/*
[INFO] Exclude: **/package-list
[INFO] Exclude: **/user.avsc
[INFO] Exclude: **/test/resources/**/*.txt
[INFO] Exclude: **/test/**/.placeholder
[INFO] Exclude: .repository/**/*
[INFO] Exclude: **/nose-*.egg/**/*
[INFO] Exclude: **/.eggs/**/*
[INFO] Exclude: **/.tox/**/*
[INFO] Exclude: **/.checkstyle
[INFO] Exclude: **/.classpath
[INFO] Exclude: **/.factorypath
[INFO] Exclude: **/.project
[INFO] Exclude: **/.settings/**/*
[INFO] Exclude: **/sdks/python/apache_beam/runners/api/*.py
[INFO] 7 resources included (use -debug for more details)
[INFO] Rat check: Summary over all files. Unapproved: 0, unknown: 0, generated: 0, approved: 7 licenses.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:28 min
[INFO] Finished at: 2017-06-08T09:17:36-07:00
[INFO] Final Memory: 102M/807M
[INFO] ------------------------------------------------------------------------

seshadri-cr commented:
@jkff, please hold on. I found the reason for test case taking such a long time and fixing it. Should be done shortly, I will update once done with that.

jkff commented:
OK, thanks. Please don't squash the new commit with your changes though - I already made my changes in my local repo, so I'll cherrypick your new commit on top of that.

seshadri-cr commented:
Optimized the unit test by marking the embedded service as a static member - https://github.com/seshadri-cr/beam/blob/3c0569f2d7a57dd141225b0ba7c301a9154d0211/sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java#L104, this way it just creates a single instance of the meta store service against which all the testing is done.  

Also added a shutdown step to the meta store service -https://github.com/seshadri-cr/beam/blob/3c0569f2d7a57dd141225b0ba7c301a9154d0211/sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/EmbeddedMetastoreService.java#L105

seshadri-cr commented:
After these changes, the testing takes 11-12 seconds in my local machine.

[INFO] --- maven-surefire-plugin:2.20:test (default-test) @ beam-sdks-java-io-hcatalog ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.beam.sdk.io.hcatalog.HCatalogIOTest
[INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 11.459 s - in org.apache.beam.sdk.io.hcatalog.HCatalogIOTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0

mvn clean verify -Prelease also passes clean in my local (I tried this on a debian-linux box as well and did not see any RAT check failures)

jkff commented:
Thanks! Applied your new commit and merged.

Feel free to take a look at the final form if you want :) (e.g. diff it against the HEAD of your repository)

seshadri-cr commented:
Thanks a ton @jkff , @iemejia  & @ssisk for your reviews and inputs in getting this done. I had definitely learned a lot, this being my first contribution. Deeply appreciate your efforts and looking forward to contribute more in coming days.

iemejia commented:
Thanks @seshadri-cr this was a really valuable contribution, and I really like the way you progressed and improved it from the initial discussions in the mailing list. Nice to see this finally merged. Congrats!

shaohuifangit commented:
@seshadri-cr 
Does it support automatic creation when the hive table is not existed
